File: src/test/java/cz/jskrabal/proxy/config/ProxyConfigurationHeadersTest.java
package cz.jskrabal.proxy.config;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import org.junit.Test;
import org.junit.runner.RunWith;

import cz.jskrabal.proxy.config.enums.ConfigurationParameter;
import junitparams.JUnitParamsRunner;
import junitparams.Parameters;

/**
 * Created by janskrabal on 11/06/16.
 */
@RunWith(JUnitParamsRunner.class)
public class ProxyConfigurationHeadersTest extends AbstractConfigurationTest {
	private static final Map<String, String> CUSTOM_HEADERS;

	static {
		Map<String, String> headers = new HashMap<>();
		headers.put("Via", "VertxProxy");
		headers.put("X-Test-Header", "test value");
		CUSTOM_HEADERS = Collections.unmodifiableMap(headers);
	}

	@SuppressWarnings("unused")
	private static Object[] provideDefaultValueTestParams() {
		return new Object[] {
				new Object[] { ConfigurationParameter.ADD_FORWARDED_BY_HEADERS },
				new Object[] { ConfigurationParameter.ADD_FORWARDED_FOR_HEADERS },
				new Object[] { ConfigurationParameter.ADD_TRANSFER_ID_HEADER },
				new Object[] { ConfigurationParameter.REMOVE_RESPONSE_HEADERS },
				new Object[] { ConfigurationParameter.REMOVE_REQUEST_HEADERS },
				new Object[] { ConfigurationParameter.CUSTOM_RESPONSE_HEADERS },
				new Object[] { ConfigurationParameter.CUSTOM_REQUEST_HEADERS }
		};
	}

	@Test
	@Parameters(method = "provideDefaultValueTestParams")
	public void testDefaultValue(ConfigurationParameter param) {
		assertDefaultValue(param);
	}

	@SuppressWarnings("unused")
	private static Object[] provideSetAndGetTestParams() {
		return new Object[] {
				new Object[] { ConfigurationParameter.ADD_FORWARDED_BY_HEADERS, true },
				new Object[] { ConfigurationParameter.ADD_FORWARDED_FOR_HEADERS, true },
				new Object[] { ConfigurationParameter.ADD_TRANSFER_ID_HEADER, true }
		};
	}

	@Test
	@Parameters(method = "provideSetAndGetTestParams")
	public void testSetAndGet(ConfigurationParameter param, Object value) {
		addValueToConfig(param, value);
		assertConfigValue(param, value);
	}

	@SuppressWarnings("unused")
	private static Object[] provideSetAndGetPojoTestParams() {
		return new Object[] {
				new Object[] { ConfigurationParameter.REMOVE_RESPONSE_HEADERS, CUSTOM_HEADERS.keySet() },
				new Object[] { ConfigurationParameter.REMOVE_REQUEST_HEADERS, CUSTOM_HEADERS.keySet() },
				new Object[] { ConfigurationParameter.CUSTOM_RESPONSE_HEADERS, CUSTOM_HEADERS },
				new Object[] { ConfigurationParameter.CUSTOM_REQUEST_HEADERS, CUSTOM_HEADERS }
		};
	}

	@Test
	@Parameters(method = "provideSetAndGetPojoTestParams")
	public void testSetAndGetPojo(ConfigurationParameter param, Object value) {
		addPojoToConfig(param, value);
		assertConfigValue(param, value);
	}

	@SuppressWarnings("unused")
	private static Object[] provideInvalidValueTestParams() {
		return new Object[] {
				new Object[] { ConfigurationParameter.ADD_FORWARDED_BY_HEADERS, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.ADD_FORWARDED_FOR_HEADERS, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.ADD_TRANSFER_ID_HEADER, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.REMOVE_RESPONSE_HEADERS, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.REMOVE_REQUEST_HEADERS, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.CUSTOM_RESPONSE_HEADERS, INVALID_NUMBER_VALUE },
				new Object[] { ConfigurationParameter.CUSTOM_REQUEST_HEADERS, INVALID_NUMBER_VALUE }
		};
	}

	@Test
	@Parameters(method = "provideInvalidValueTestParams")
	public void testInvalidValue(ConfigurationParameter param, Object invalidValue) {
		addValueToConfig(param, invalidValue);
		assertDefaultValue(param);
	}
}

File: src/test/kotlin/cz/jskrabal/proxy/config/ProxyConfigurationHeadersTest.kt
package cz.jskrabal.proxy.config

import java.util.Collections
import java.util.HashMap

import org.junit.Test
import org.junit.runner.RunWith

import cz.jskrabal.proxy.config.enums.ConfigurationParameter
import junitparams.JUnitParamsRunner
import junitparams.Parameters

/**
 * Created by janskrabal on 11/06/16.
 */
@RunWith(JUnitParamsRunner::class)
class ProxyConfigurationHeadersTest : AbstractConfigurationTest() {

    @Test
    @Parameters(method = "provideDefaultValueTestParams")
    fun testDefaultValue(param: ConfigurationParameter) {
        assertDefaultValue(param)
    }

    @Test
    @Parameters(method = "provideSetAndGetTestParams")
    fun testSetAndGet(param: ConfigurationParameter, value: Any) {
        addValueToConfig(param, value)
        assertConfigValue(param, value)
    }

    @Test
    @Parameters(method = "provideSetAndGetPojoTestParams")
    fun testSetAndGetPojo(param: ConfigurationParameter, value: Any) {
        addPojoToConfig(param, value)
        assertConfigValue(param, value)
    }

    @Test
    @Parameters(method = "provideInvalidValueTestParams")
    fun testInvalidValue(param: ConfigurationParameter, invalidValue: Any) {
        addValueToConfig(param, invalidValue)
        assertDefaultValue(param)
    }

    companion object {
        private val CUSTOM_HEADERS: Map<String, String>

        init {
            val headers = HashMap<String, String>()
            headers.put("Via", "VertxProxy")
            headers.put("X-Test-Header", "test value")
            CUSTOM_HEADERS = Collections.unmodifiableMap(headers)
        }

        @JvmStatic
        private fun provideDefaultValueTestParams(): Array<Any> {
            return arrayOf(arrayOf<Any>(ConfigurationParameter.ADD_FORWARDED_BY_HEADERS), arrayOf<Any>(ConfigurationParameter.ADD_FORWARDED_FOR_HEADERS), arrayOf<Any>(ConfigurationParameter.ADD_TRANSFER_ID_HEADER), arrayOf<Any>(ConfigurationParameter.REMOVE_RESPONSE_HEADERS), arrayOf<Any>(ConfigurationParameter.REMOVE_REQUEST_HEADERS), arrayOf<Any>(ConfigurationParameter.CUSTOM_RESPONSE_HEADERS), arrayOf<Any>(ConfigurationParameter.CUSTOM_REQUEST_HEADERS))
        }

        @JvmStatic
        private fun provideSetAndGetTestParams(): Array<Any> {
            return arrayOf(arrayOf(ConfigurationParameter.ADD_FORWARDED_BY_HEADERS, true), arrayOf(ConfigurationParameter.ADD_FORWARDED_FOR_HEADERS, true), arrayOf(ConfigurationParameter.ADD_TRANSFER_ID_HEADER, true))
        }

        @JvmStatic
        private fun provideSetAndGetPojoTestParams(): Array<Any> {
            return arrayOf(arrayOf(ConfigurationParameter.REMOVE_RESPONSE_HEADERS, CUSTOM_HEADERS.keys), arrayOf(ConfigurationParameter.REMOVE_REQUEST_HEADERS, CUSTOM_HEADERS.keys), arrayOf(ConfigurationParameter.CUSTOM_RESPONSE_HEADERS, CUSTOM_HEADERS), arrayOf(ConfigurationParameter.CUSTOM_REQUEST_HEADERS, CUSTOM_HEADERS))
        }

        @JvmStatic
        private fun provideInvalidValueTestParams(): Array<Any> {
            return arrayOf(arrayOf(ConfigurationParameter.ADD_FORWARDED_BY_HEADERS, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.ADD_FORWARDED_FOR_HEADERS, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.ADD_TRANSFER_ID_HEADER, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.REMOVE_RESPONSE_HEADERS, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.REMOVE_REQUEST_HEADERS, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.CUSTOM_RESPONSE_HEADERS, AbstractConfigurationTest.INVALID_NUMBER_VALUE), arrayOf(ConfigurationParameter.CUSTOM_REQUEST_HEADERS, AbstractConfigurationTest.INVALID_NUMBER_VALUE))
        }
    }
}
